File: Templates/ZIKRouter.xctemplate/ZIKServiceRoutermodule inputObjective-C/___FILEBASENAME___.m
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//___COPYRIGHT___
//

#import "___VARIABLE_productName___.h"
#import <ZIKRouter/ZIKRouterInternal.h>
#import <ZIKRouter/ZIKServiceRouterInternal.h>
#import "___VARIABLE_destinationClass___.h"

DeclareRoutableService(___VARIABLE_destinationClass___, ___VARIABLE_productName___)
DeclareRoutableServiceModuleProtocol(___VARIABLE_moduleProtocolName___)

@interface ___VARIABLE_productName___ ()

@end

@implementation ___VARIABLE_productName___

+ (void)registerRoutableDestination {
    [self registerExclusiveService:[___VARIABLE_destinationClass___ class]];
    [self registerModuleProtocol:ZIKRoutable(___VARIABLE_moduleProtocolName___)];
}

// Return your custom configuration
+ (ZIKServiceMakeableConfiguration<___VARIABLE_moduleProtocolName___> *)defaultRouteConfiguration {
    ZIKServiceMakeableConfiguration<id<___VARIABLE_protocolName___>> *config = [[ZIKServiceMakeableConfiguration alloc] init];
    __weak typeof(config) weakConfig = config;
    
    config._prepareDestination = ^(id<___VARIABLE_protocolName___> destination) {
        // Prepare destination
    };
    
    config.makeDestinationWith = ^id<___VARIABLE_protocolName___> _Nullable(/*arguments*/) {
        weakConfig.makeDestination = ^id<___VARIABLE_protocolName___> _Nullable{
#error TODO: instantiate destination
            // Instantiate destination. Return nil if configuration is invalid.
            ___VARIABLE_destinationClass___ *destination = [[___VARIABLE_destinationClass___ alloc] init];
            return destination;
        };
        weakConfig.makedDestination = weakConfig.makeDestination();
        if (weakConfig._prepareDestination) {
            weakConfig._prepareDestination(weakConfig.makedDestination);
        }
        return weakConfig.makedDestination;
    };
    return config;
}

- (nullable ___VARIABLE_destinationClass___ *)destinationWithConfiguration:(ZIKServiceMakeableConfiguration<___VARIABLE_moduleProtocolName___> *)configuration {
    if (configuration.makeDestination) {
        return configuration.makeDestination();
    }
#error TODO: instantiate destination
    // Instantiate destination with configuration. Return nil if configuration is invalid.
    ___VARIABLE_destinationClass___ *destination = [[___VARIABLE_destinationClass___ alloc] init];

    return destination;
}

@end

File: Templates/ZIKRouter.xctemplate/ZIKServiceRoutermodule inputSwift/___FILEBASENAME___.swift
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//___COPYRIGHT___
//

import ZRouter
import ZIKRouter.Internal

class ___VARIABLE_productName___: ZIKServiceRouter<___VARIABLE_destinationClass___, PerformRouteConfig> {
    
    override class func registerRoutableDestination() {
        registerExclusiveService(___VARIABLE_destinationClass___.self)
        register(RoutableServiceModule<___VARIABLE_moduleProtocolName___>())
    }
    
    override class func defaultRouteConfiguration() -> PerformRouteConfig {
        let config = ServiceMakeableConfiguration<___VARIABLE_protocolName___, (/*arguments*/) -> ___VARIABLE_protocolName___?>({ _ in return nil })
        
        config.__prepareDestination = { destination in
            // Prepare destination
        }
        
        config.makeDestinationWith = { [unowned config] (arguments) in
            config.makeDestination = { () in
                // Instantiate destination. Return nil if configuration is invalid.
                let destination: ___VARIABLE_destinationClass___? = /*___VARIABLE_destinationClass___()*/
                return destination
            }
            if let destination = config.makeDestination?() {
                config.__prepareDestination?(destination)
                config.makedDestination = destination
                return destination
            }
            return nil
        }
        return config
    }
    
    override func destination(with configuration: PerformRouteConfig) -> ___VARIABLE_destinationClass___? {
        if let config = configuration as? ServiceMakeableConfiguration<___VARIABLE_protocolName___, (/*arguments*/) -> ___VARIABLE_protocolName___?>,
            let makeDestination = config.makeDestination {
            return makeDestination() as? ___VARIABLE_destinationClass___ ?? nil
        }
        // Instantiate destination with configuration. Return nil if configuration is invalid.
        let destination: ___VARIABLE_destinationClass___? = /*___VARIABLE_destinationClass___()*/
        return destination
    }
    
}

extension ___VARIABLE_destinationClass___: ZIKRoutableService {
    
}

extension RoutableServiceModule where Protocol == ___VARIABLE_moduleProtocolName___ {
    init() { self.init(declaredTypeName: "___VARIABLE_moduleProtocolName___") }
}

extension ServiceMakeableConfiguration: ___VARIABLE_moduleProtocolName___ where Destination == ___VARIABLE_protocolName___, Constructor == (/*arguments*/) -> ___VARIABLE_protocolName___? {
    
}
